src/vr/journal.zig

```diff
@@ -489,129 +729,241 @@ pub const Journal = struct {
         };
     }
 
-    /// Writes headers between `op_min` and `op_max` (both inclusive).
-    fn write_headers_between(self: *Journal, op_min: u64, op_max: u64) void {
-        // TODO Snapshots
-        assert(op_min <= op_max);
+    /// Attempt to lock the in-memory sector containing the header being written.
+    /// If the sector is already locked, add this write to the wait queue.
+    fn write_prepare_header(self: *Journal, write: *Journal.Write) void {
+        assert(!write.header_sector_locked);
+        assert(write.header_sector_next == null);
+
+        var it = self.writes.iterate();
+        while (it.next()) |other| {
+            if (other == write) continue;
+            if (!other.header_sector_locked) continue;
+
+            if (other.header_sector_same(write)) {
+                write.header_sector_next = other.header_sector_next;
+                other.header_sector_next = write;
+                return;
+            }
+        }
 
-        const offset = Journal.sector_floor(op_min * @sizeOf(Header));
-        const len = Journal.sector_ceil((op_max - op_min + 1) * @sizeOf(Header));
-        assert(len > 0);
+        write.header_sector_locked = true;
+        self.write_prepare_on_lock_header_sector(write);
```

Comment:
Let's move the function definition for `self.write_prepare_on_lock_header_sector()` directly after the end of this function where we call it?